consensus/core/src/commit_finalizer.rs

```diff
@@ -605,15 +608,40 @@ impl CommitFinalizer {
                 continue;
             }
             let curr_block_state = blocks_map.get(&curr_block_ref).unwrap_or_else(|| panic!("Block {curr_block_ref} is either incorrectly gc'ed or failed to be recovered after crash.")).read();
-            // Ignore info from the block if its direct ancestor has been processed.
+            // The first ancestor of current block should have the same origin / author as the current block.
+            // If it is not found in the blocks map but have round higher than the pending block, it might have
+            // voted on the pending block but have been GC'ed.
+            // Because the GC'ed block might have voted on the pending block and rejected some of the pending transactions,
+            // we cannot assume current block is voting to accept transactions from the pending block.
+            let curr_origin_ancestor_ref = curr_block_state.block.ancestors().first().unwrap();
+            let skip_votes = curr_block_ref.author == curr_origin_ancestor_ref.author
+                && pending_block_ref.round < curr_origin_ancestor_ref.round
+                && !blocks_map.contains_key(curr_origin_ancestor_ref);
```

Comment:
I am still wondering if we need to do more here. It seems confusing to me even in this rare case that we would not count all the votes properly. Correct me if I am wrong, but it looks like we skip here out of caution that there could be a GC block even though there is a chance the vote could be counted. Would we not be able to track each blocks direct descendant aka their voting block per authority in BlockState. This way we would have concrete evidence of which blocks to be used for votes.